### PR TITLE
Provide breadcrumbs

### PR DESCRIPTION
**Pull request by jasonrhodes**
WIP to replace `react-router-breadcrumbs-hoc` with our own implementation to take control of types and making sure breadcrumbs are returned as strings. 

Uses a render prop over the `withBreadcrumbs` HOC, which made types a lot easier.

### CODE REVIEW

File: x-pack/plugins/apm/public/components/app/Main/routeConfig.tsx [hunk 1]
```diff
@@ -64,49 +90,6 @@ export const routes: Route[] = [
       defaultMessage: 'Errors'
     })
   },
-  {
```
Comment (jasonrhodes) [inline x-pack/plugins/apm/public/components/app/Main/routeConfig.tsx]: If we wrap all of our main routes in one `<Switch>`, we don't need to worry about the "switchRoutes" logic and can simplify the routes config, and no longer need to "flatten" the routes before giving the routes config to the breadcrumbs component, etc.

File: x-pack/plugins/apm/public/components/app/Main/routeConfig.tsx [hunk 2]
```diff
@@ -43,18 +30,57 @@ const renderAsRedirectTo = (to: string) => {
...
 
-export const routes: Route[] = [
+export const routes: BreadcrumbRoute[] = [
   {
     exact: true,
     path: '/',
     render: renderAsRedirectTo('/services'),
     breadcrumb: 'APM'
   },
+  {
```
Comment (jasonrhodes) [inline x-pack/plugins/apm/public/components/app/Main/routeConfig.tsx]: If we wrap all of our main routes in one <Switch>, we don't need to worry about the "switchRoutes" logic and can simplify the routes config here, and no longer need to "flatten" the routes before giving the routes config to the breadcrumbs component, etc.

File: x-pack/plugins/apm/public/components/app/Main/index.js [hunk 3]
```diff
@@ -25,17 +25,11 @@ export default function Main() {
       <UpdateBreadcrumbs />
       <Route component={ConnectRouterToRedux} />
       <Route component={ScrollToTopOnPathChange} />
-      {routes.map((route, i) => {
```
Comment (jasonrhodes) [inline x-pack/plugins/apm/public/components/app/Main/index.js]: See comment below about wrapping the main routes in a single `<Switch>`

**Comment by sorenlouv** [comment]
@jasonrhodes Any reason not to open this against master? `dynamic-title` and `ProvideBreadcrumbs` feels very intertwined, so hard to review them separately.

**Comment by jasonrhodes** [comment]
@sqren I wanted to get thoughts/opinions on this approach (pulling in the breadcrumb code as a fork) separate from the other work in dynamic-title...

**Comment by jasonrhodes** [comment]
Created this to give you options https://github.com/elastic/kibana/pull/29960

**Comment by jasonrhodes** [comment]
Closed in favor of elastic#29960

**Pull request closed without merging**